flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -334,8 +335,11 @@ public void onContainersCompleted(final List<ContainerStatus> list) {
 					if (yarnWorkerNode != null) {
 						// Container completed unexpectedly ~> start a new one
 						final Container container = yarnWorkerNode.getContainer();
-						requestYarnContainer(container.getResource(), yarnWorkerNode.getContainer().getPriority());
-						closeTaskManagerConnection(resourceId, new Exception(containerStatus.getDiagnostics()));
+						// check WorkerRegistration status to avoid requesting containers more than required
+						if (checkWorkerRegistrationWithResourceId(resourceId)) {
```

Comment:
Wouldn't that prevent container restarts if the container failure happened before the `TaskManager` registered at the `ResourceManager`, because then, `ResourceManager#taskExecutors` would not contain the given `ResourceID`?

Reply:
Yes, I might happen. The problem is not as easy as I thought. The actual cause of this problem is the resource was released before a full restart but the onContainerCompleted callback method happened after the full restart. As the full restart will requesting all the containers needed as configured, if the onContainerCompleted  method was called after that, it will request for a new container and possess it which is not needed.

Reply:
Yes, I think it is not possible to distinguish between a container which was released but has not been completed before a recovery and a container failure just after recovery without some kind of state (in both cases we retrieve the containers from the previous attempt and get a onContainerCompleted signal from the container).

My question would be how often does it happen that we run into this situation. Releasing a container and failing immediately afterwards should happen fairly rarely, I would assume. Moreover, at some point the an idle `TaskManager` should be released and, thus, also the underlying container.

Reply:
What we maybe could do instead of counting simply how many pending container requests we have is to ask the `SlotManager` how many pending slot allocations it has. If the number of slot allocations is lower than the product of slots per task manager x pending container requests, then we would not have to restart the container.

Reply:
Yes, you are right. I have been thinking about how to add a global version state to a container all the week, but feels like I came into a dead end. I found this question when there are higher priority batch job in yarn cluster which occupied all the resources and made yarn kill the flink task manager containers frequently in a period of time.

Reply:
I will work on the implementation of the idea you mentioned above which makes perfect sense to me.